homeassistant/components/camera/__init__.py

```diff
@@ -0,0 +1,1133 @@
+                camera.is_streaming = False
+                camera.update_ha_state()
+
+            message = "{0} stopped streaming to {1}".format(
+                camera.name,
+                handler.address_string())
+
+            hass.bus.fire(
+                "camera_stream_stopped",
+                {
+                    "component": DOMAIN,
+                    ATTR_ENTITY_ID: entity_id,
+                    ATTR_FRIENDLY_LOG_MESSAGE: message})
+
+        else:
+            handler.send_response(HTTP_NOT_FOUND)
+
+        camera.is_streaming = False
+
+    hass.http.register_path(
+        'GET',
+        re.compile(
+            r'/api/camera_proxy_stream/(?P<entity_id>[a-zA-Z\._0-9]+)'),
+        _proxy_camera_mjpeg_stream,
+        require_auth=True)
+
+    def _get_camera_recordings(handler, path_match, data):
+        """ Returns a JSON object with a list of camera recording such as still
+        image snapshots, motion detection recordings and recordings. """
+        entity_id = path_match.group('entity_id')
+
+        camera = None
+        if entity_id in component.entities.keys():
+            camera = component.entities[entity_id]
+
+        if camera:
+            offset = int(data.get('offset', 0))
+            length = int(data.get('length', 10))
+            rec_type = data.get('type', 'snapshot')
+
+            rec_path = None
+            if rec_type == 'snapshot':
+                rec_path = camera.snapshot_images_path
+            elif rec_type == 'recording':
+                rec_path = camera.recording_images_path
+            elif rec_type == 'motion':
+                rec_path = camera.event_images_path
+
+            data = camera.get_all_recordings(rec_path, offset, length)
+            if len(data) == 0:
+                handler.send_response(HTTP_NOT_FOUND)
+                handler.end_headers()
+            else:
+                handler.write_json(data)
+        else:
+            handler.send_response(HTTP_NOT_FOUND)
+            handler.end_headers()
+
+    hass.http.register_path(
+        'GET',
+        re.compile(r'/api/camera_recordings/(?P<entity_id>[a-zA-Z\._0-9]+)'),
+        _get_camera_recordings,
+        require_auth=True)
+
+    def _saved_camera_image(handler, path_match, data):
+        """ This creates an API endpoint that serves a saved camera image from
+        disk.  It also does some basic checking to make sure relative paths
+        outside of the legitimate locations cn't be used """
+        entity_id = path_match.group('entity_id')
+
+        camera = None
+        if entity_id in component.entities.keys():
+            camera = component.entities[entity_id]
+
+        if camera:
+            image_path = os.path.normpath(os.path.join(
+                camera.images_path, data['image_path']))
+
+            # Check to see that someone is not trying to do something dodgey
+            # with relative paths
+            if not image_path.startswith(camera.images_path):
```

Comment:
Besides this check also please use `homeassistant.util.santize_path`. It will remove .. and ~